File: research/experiments/sweep/runs/opt-small-swift-sonnet45/Sources/Optimization/ResultTypes.swift
import Foundation

/// Optimization options with convergence tolerances
public struct OptimizeOptions {
    public let gradTol: Double
    public let stepTol: Double
    public let funcTol: Double
    public let maxIterations: Int

    public init(
        gradTol: Double = 1e-8,
        stepTol: Double = 1e-8,
        funcTol: Double = 1e-12,
        maxIterations: Int = 1000
    ) {
        self.gradTol = gradTol
        self.stepTol = stepTol
        self.funcTol = funcTol
        self.maxIterations = maxIterations
    }
}

/// Convergence reason (tagged union)
public enum ConvergenceReason: Equatable {
    case gradient
    case step
    case function
    case maxIterations
    case lineSearchFailed
}

/// Optimization result
public struct OptimizeResult {
    public let x: [Double]
    public let fun: Double
    public let gradient: [Double]?
    public let iterations: Int
    public let functionCalls: Int
    public let gradientCalls: Int
    public let converged: Bool
    public let message: String

    public init(
        x: [Double],
        fun: Double,
        gradient: [Double]?,
        iterations: Int,
        functionCalls: Int,
        gradientCalls: Int,
        converged: Bool,
        message: String
    ) {
        self.x = x
        self.fun = fun
        self.gradient = gradient
        self.iterations = iterations
        self.functionCalls = functionCalls
        self.gradientCalls = gradientCalls
        self.converged = converged
        self.message = message
    }
}

/// Result type functions
public enum ResultTypes {

    /// Create default options with optional overrides
    public static func defaultOptions(
        gradTol: Double? = nil,
        stepTol: Double? = nil,
        funcTol: Double? = nil,
        maxIterations: Int? = nil
    ) -> OptimizeOptions {
        return OptimizeOptions(
            gradTol: gradTol ?? 1e-8,
            stepTol: stepTol ?? 1e-8,
            funcTol: funcTol ?? 1e-12,
            maxIterations: maxIterations ?? 1000
        )
    }

    /// Check convergence criteria in order: gradient → step → function → maxIterations
    public static func checkConvergence(
        gradNorm: Double,
        stepNorm: Double,
        funcChange: Double,
        iteration: Int,
        options: OptimizeOptions
    ) -> ConvergenceReason? {
        if gradNorm < options.gradTol {
            return .gradient
        }
        if stepNorm < options.stepTol {
            return .step
        }
        if funcChange < options.funcTol {
            return .function
        }
        if iteration >= options.maxIterations {
            return .maxIterations
        }
        return nil
    }

    /// Check if a convergence reason represents true convergence
    public static func isConverged(_ reason: ConvergenceReason) -> Bool {
        switch reason {
        case .gradient, .step, .function:
            return true
        case .maxIterations, .lineSearchFailed:
            return false
        }
    }

    /// Human-readable convergence message
    public static func convergenceMessage(_ reason: ConvergenceReason) -> String {
        switch reason {
        case .gradient:
            return "Converged: gradient norm below tolerance"
        case .step:
            return "Converged: step size below tolerance"
        case .function:
            return "Converged: function change below tolerance"
        case .maxIterations:
            return "Stopped: maximum iterations reached"
        case .lineSearchFailed:
            return "Stopped: line search failed"
        }
    }
}

File: research/experiments/sweep/runs/opt-small-csharp-sonnet45/Optimization/ResultTypes.cs
namespace Optimization;

/// <summary>
/// Options for optimization algorithms.
/// </summary>
public class OptimizeOptions
{
    public double GradTol { get; set; } = 1e-8;
    public double StepTol { get; set; } = 1e-8;
    public double FuncTol { get; set; } = 1e-12;
    public int MaxIterations { get; set; } = 1000;
}

/// <summary>
/// Result of an optimization run.
/// </summary>
public class OptimizeResult
{
    public required double[] X { get; init; }
    public required double Fun { get; init; }
    public double[]? Gradient { get; init; }
    public required int Iterations { get; init; }
    public required int FunctionCalls { get; init; }
    public required int GradientCalls { get; init; }
    public required bool Converged { get; init; }
    public required string Message { get; init; }
}

/// <summary>
/// Tagged union representing convergence reasons.
/// </summary>
public abstract record ConvergenceReason
{
    public sealed record Gradient : ConvergenceReason;
    public sealed record Step : ConvergenceReason;
    public sealed record Function : ConvergenceReason;
    public sealed record MaxIterations : ConvergenceReason;
    public sealed record LineSearchFailed : ConvergenceReason;
}

/// <summary>
/// Functions for working with optimization options and convergence.
/// </summary>
public static class ResultTypesHelper
{
    /// <summary>
    /// Create default options with optional overrides.
    /// </summary>
    public static OptimizeOptions DefaultOptions(Action<OptimizeOptions>? configure = null)
    {
        var options = new OptimizeOptions();
        configure?.Invoke(options);
        return options;
    }

    /// <summary>
    /// Check convergence criteria in order: gradient → step → function → maxIterations.
    /// Returns null if no criterion is met.
    /// </summary>
    public static ConvergenceReason? CheckConvergence(
        double gradNorm,
        double stepNorm,
        double funcChange,
        int iteration,
        OptimizeOptions opts)
    {
        if (gradNorm < opts.GradTol)
            return new ConvergenceReason.Gradient();
        if (stepNorm < opts.StepTol)
            return new ConvergenceReason.Step();
        if (funcChange < opts.FuncTol)
            return new ConvergenceReason.Function();
        if (iteration >= opts.MaxIterations)
            return new ConvergenceReason.MaxIterations();
        return null;
    }

    /// <summary>
    /// Check if a convergence reason indicates successful convergence.
    /// </summary>
    public static bool IsConverged(ConvergenceReason reason)
    {
        return reason switch
        {
            ConvergenceReason.Gradient => true,
            ConvergenceReason.Step => true,
            ConvergenceReason.Function => true,
            ConvergenceReason.MaxIterations => false,
            ConvergenceReason.LineSearchFailed => false,
            _ => false
        };
    }

    /// <summary>
    /// Get a human-readable message for a convergence reason.
    /// </summary>
    public static string ConvergenceMessage(ConvergenceReason reason)
    {
        return reason switch
        {
            ConvergenceReason.Gradient => "Converged: gradient norm below tolerance",
            ConvergenceReason.Step => "Converged: step size below tolerance",
            ConvergenceReason.Function => "Converged: function change below tolerance",
            ConvergenceReason.MaxIterations => "Stopped: maximum iterations reached",
            ConvergenceReason.LineSearchFailed => "Stopped: line search failed",
            _ => "Unknown convergence reason"
        };
    }
}
